src/EntityComponentManager.cc

```diff
@@ -270,6 +276,154 @@ Entity EntityComponentManagerPrivate::CreateEntityImplementation(Entity _entity)
   return _entity;
 }
 
+/////////////////////////////////////////////////
+Entity EntityComponentManager::Clone(const Entity _entity, const Entity _parent,
+    const std::string &_name, bool _allowRename)
+{
+  // During cloning, we populate two maps:
+  //  - map of cloned model entities to the non-cloned model's canonical link
```

Comment:
If feels strange to give the canonical link component special treatment. The ECM should be treating all components the same way as much as possible.

I believe that a more general approach would be to treat all components that have `Entity` data type the same way. This may need some template magic. At the moment, the only components that hold `Entity` are `CanonicalLink` and `ParentEntity`, but in the future we may have other components that point to other entities (I don't know, like `EndEffectorLink`), and the current approach won't scale well for them.

Reply:
> If feels strange to give the canonical link component special treatment. The ECM should be treating all components the same way as much as possible.

Agreed :+1: 

> I believe that a more general approach would be to treat all components that have Entity data type the same way. This may need some template magic.

Should I try to go ahead and update this now? Or do we want to leave a `TODO` note/ticket an issue and come back to it later? If we were to make this change now, I'm not sure off the top of my head how I'd do it, so we may need to do some brainstorming.

Reply:
I'd be ok going with this for now for expediency, but how about removing that private function from the public header?

Reply:
> how about removing that private function from the public header?

I was hoping to move the private method to the `EntityComponentManagerPrivate` class, but I can't because the private method calls private methods from the `EntityComponentManager` class (`CreateComponentImplementation`, `ComponentImplementation`). So, unless there's another way to hide this method, it may need to stay in the public header.

Reply:
I pushed the idea we talked about in 96396c83c5947151ea3050bbd8234d89eccc3118.